### PR TITLE
[CDAP-13985] Upgrade Babel + Webpack & add the ability to write typescript for UI development

### PR DESCRIPTION
**Pull request by ajainarayanan**
JIRA: https://issues.cask.co/browse/CDAP-13985
Build: https://builds.cask.co/browse/CDAP-UDUT24

**TL;DR -**
- Upgrades webpack and babel. This is required for us to be able to use latest version of typescript.
- Adds typescript support. So now any new component we write can be written in typescript. We can start enforcing types in UI for better stability and predicatability
- Adds set of sample components as way to demonstrate how we might start writing components in typescript.
- Adds `clean-webpack-plugin` that nukes the dist folders before starting the build. This will prevent the dist folders growing over the period of development.

**Note:** 
- Right now merging to the react-upgrade branch as this is branched off from that branch. Will change the base branch if the react upgrade branch is merged before this review completes.

- Babel is now deprecating the use of stage presets. For [reference](https://babeljs.io/blog/2018/07/27/removing-babels-stage-presets). When we upgrade beyond `beta.54` we need to remove the presets from `.babelrc` and replace it with appropriate plugins 😭 

- Typescript released 3.0 version of typescript. We don't have to jump to the latest version but this is something we might do in another 3-6 months time 😭 😭. For [reference ](https://blogs.msdn.microsoft.com/typescript/2018/07/30/announcing-typescript-3-0/)

- Have disabled almost all the js rules in tslint that requires re-formatting/re-writing existing js apps. Will add these back gradually when we add prettier to the build process.

- We can still do better in our code splitting. Right now there are files that are just too big. We need to spend sometime in figuring out a better split strategy.

### CODE REVIEW

**Review by elfenheart — APPROVED** [review]
👍  LGTM
